### PR TITLE
fix: fix grant and revoke privileges for internal table

### PR DESCRIPTION
**Pull request by yezizp2012**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As title.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by kwannoel — APPROVED** [review]
After this PR, we will grant permissions on internal tables of MVs, when we grant permissions on the MV itself.

Can we also extend this to sinks and indexes?

**Comment by yezizp2012** [comment]
> After this PR, we will grant permissions on internal tables of MVs, when we grant permissions on the MV itself.
> 
> Can we also extend this to sinks and indexes?

👍 Yes sure, I will extend it in this PR later.

**Comment by yezizp2012** [comment]
> After this PR, we will grant permissions on internal tables of MVs, when we grant permissions on the MV itself.
> 
> Can we also extend this to sinks and indexes?

After this PR, The `select` privilege on an index will be granted or revoked correspondingly with its primary table. This PR also supports grant and revoke `select` privilege for sink, which means user with `select` privilege on sink can query its internal tables now. Cc @kwannoel PTAL.

File: src/common/src/acl/mod.rs [hunk 1]
```diff
@@ -104,7 +104,7 @@ pub static ALL_AVAILABLE_TABLE_MODES: LazyLock<AclModeSet> =
     LazyLock::new(|| make_bitflags!(AclMode::{Select | Insert | Update | Delete}).into());
 pub static ALL_AVAILABLE_SOURCE_MODES: LazyLock<AclModeSet> = LazyLock::new(AclModeSet::readonly);
 pub static ALL_AVAILABLE_MVIEW_MODES: LazyLock<AclModeSet> = LazyLock::new(AclModeSet::readonly);
-pub static ALL_AVAILABLE_SINK_MODES: LazyLock<AclModeSet> = LazyLock::new(AclModeSet::empty);
+pub static ALL_AVAILABLE_SINK_MODES: LazyLock<AclModeSet> = LazyLock::new(AclModeSet::readonly);
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply src/common/src/acl/mod.rs:107]: Nevermind they are just risingwave's own Access controls.

File: src/frontend/src/handler/handle_privilege.rs [hunk 2]
```diff
@@ -136,6 +136,20 @@ fn make_prost_privilege(
...
             }
         }
+        GrantObjects::Sinks(sinks) => {
+            let db_name = session.database();
+            let search_path = session.config().search_path();
+            let user_name = &session.auth_context().user_name;
+
+            for name in sinks {
+                let (schema_name, sink_name) =
+                    Binder::resolve_schema_qualified_name(db_name, name)?;
```
Comment (kwannoel) [inline src/frontend/src/handler/handle_privilege.rs:146]: I assume we just resolve the schema based on the session schema here?

File: src/meta/src/controller/utils.rs [hunk 3]
```diff
@@ -735,6 +736,58 @@ pub fn construct_privilege_dependency_query(ids: Vec<PrivilegeId>) -> WithQuery
...
+                    .and(table::Column::BelongsToJobId.is_in(index_table_ids.clone())),
+            )
+            .into_tuple()
+            .all(db)
+            .await?;
+
+        index_table_ids.extend(internal_table_ids.into_iter());
+    }
+
+    Ok(index_table_ids)
```
Comment (kwannoel) [inline src/meta/src/controller/utils.rs:788]: Why do we need to return `internal_tables` as well, shouldn't `get_internal_tables_by_id` called on the index_table_id handle that instead? 🤔

File: src/meta/src/controller/user.rs [hunk 4]
```diff
@@ -351,12 +369,29 @@ impl CatalogController {
...
+                    if action == Action::Select {
+                        include_select = true;
+                    }
+                    action
+                })
                 .collect_vec();
             revoke_items.insert(obj, actions);
+            if include_select {
+                revoke_items.extend(
+                    internal_table_ids
```
Comment (kwannoel) [inline src/meta/src/controller/user.rs:389]: Is it expected that only `Select` requires revoking privileges on `internal_tables`?

File: src/frontend/src/handler/handle_privilege.rs (hunk 2 above)
Comment (yezizp2012) [reply src/frontend/src/handler/handle_privilege.rs:146]: Nope, it only resolve the schema if user specified it. For example: `GRANT SELECT ON SINK schema.sink_name TO user`. If user doesn't specify it, the schema will be resolved by the session search path later.

File: src/meta/src/controller/utils.rs (hunk 3 above)
Comment (yezizp2012) [reply src/meta/src/controller/utils.rs:788]: We don't support grant or revoke privileges on index directly, here the logic is: if user want to grant or revoke `SELECT` privilege on a table, its internal tables, index tables and internal tables of indexes will be grant or revoke `SELECT` correspondingly.

File: src/meta/src/controller/user.rs (hunk 4 above)
Comment (yezizp2012) [reply src/meta/src/controller/user.rs:389]: Yes, we only have `SELECT` privilege for internal tables.

**Review by kwannoel — APPROVED** [review]
LGTM, thanks for the work.

**Pull request merged**